File: node/src/manifest/manifest.js
import { readFile } from 'node:fs/promises';
import path from 'node:path';

const cache = new Map();

export function clearManifestCache() {
  cache.clear();
}

export async function load(coldStorageRoot, tenantId) {
  if (cache.has(tenantId)) return cache.get(tenantId);

  const manifestPath = path.join(coldStorageRoot, tenantId, 'manifest.json');
  const data = JSON.parse(await readFile(manifestPath, 'utf-8'));
  cache.set(tenantId, data);
  return data;
}

export function filesInRange(manifest, startTs, endTs) {
  return manifest.files.filter((f) => f.endTs >= startTs && f.startTs <= endTs);
}

File: go/internal/manifest/manifest.go
package manifest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sync"

	"ducker/internal/model"
)

var (
	mu    sync.RWMutex
	cache = map[string]model.Manifest{}
)

func cacheKey(root, tenantID string) string {
	return root + "::" + tenantID
}

func Clear() {
	mu.Lock()
	defer mu.Unlock()
	cache = map[string]model.Manifest{}
}

func Load(coldStorageRoot, tenantID string) (model.Manifest, error) {
	key := cacheKey(coldStorageRoot, tenantID)
	mu.RLock()
	if m, ok := cache[key]; ok {
		mu.RUnlock()
		return m, nil
	}
	mu.RUnlock()

	manifestPath := filepath.Join(coldStorageRoot, tenantID, "manifest.json")
	data, err := os.ReadFile(manifestPath)
	if err != nil {
		return model.Manifest{}, err
	}

	var m model.Manifest
	if err := json.Unmarshal(data, &m); err != nil {
		return model.Manifest{}, err
	}

	mu.Lock()
	cache[key] = m
	mu.Unlock()
	return m, nil
}

func FilesInRange(m model.Manifest, startTS, endTS int64) []model.SegmentFile {
	out := make([]model.SegmentFile, 0, len(m.Files))
	for _, f := range m.Files {
		if f.EndTS >= startTS && f.StartTS <= endTS {
			out = append(out, f)
		}
	}
	return out
}
